app/src/main/kotlin/io/homeassistant/companion/android/webview/WebViewPresenterImpl.kt

```diff
@@ -85,68 +87,180 @@ class WebViewPresenterImpl @Inject constructor(
+     */
+    private suspend fun ensureServerIsAvailable(): Boolean {
+        val previousServerId = serverId
+        serverManager.getServer(serverId) ?: setActiveServer(ServerManager.SERVER_ID_ACTIVE)
+        return serverId != previousServerId
+    }
+
+    /**
+     * Checks if the current server has an authenticated session.
+     *
+     * @return `true` if the session is connected, `false` if anonymous or unavailable
+     */
+    private suspend fun isSessionConnected(): Boolean {
+        return try {
+            serverManager.authenticationRepository(serverId).getSessionState() != SessionState.ANONYMOUS
+        } catch (e: IllegalArgumentException) {
+            Timber.w(e, "Unable to get server session state, not continuing")
+            false
+        }
+    }
 
-            /*
-            We only want to cause the UI to reload if the server or URL that we need to load has
-            changed. An example of this would be opening the app on wifi with a local url then
-            loosing wifi signal and reopening app. Without this we would still be trying to use the
-            internal url externally.
-             */
-            if (
-                oldUrlForServer != urlForServer ||
-                oldUrl?.protocol != url?.protocol ||
-                oldUrl?.host != url?.host ||
-                oldUrl?.port != url?.port
-            ) {
-                view.loadUrl(
-                    url = url.toString().toUri()
-                        .buildUpon()
-                        .appendQueryParameter("external_auth", "1")
-                        .build()
-                        .toString(),
-                    keepHistory = oldUrlForServer == urlForServer,
-                    openInApp = url?.baseIsEqual(baseUrl) ?: false,
-                )
-            }
+    private suspend fun handleUrlState(
+        urlState: UrlState,
+        path: String?,
+        shouldConsumePath: Boolean,
+        isNewServer: Boolean,
+    ) {
+        when (urlState) {
+            is UrlState.HasUrl -> handleHasUrlState(
+                urlState = urlState,
+                path = path,
+                shouldConsumePath = shouldConsumePath,
+                isNewServer = isNewServer,
+            )
+            UrlState.InsecureState -> handleInsecureState()
         }
     }
 
+    private suspend fun handleHasUrlState(
+        urlState: UrlState.HasUrl,
+        path: String?,
+        shouldConsumePath: Boolean,
+        isNewServer: Boolean,
+    ) {
+        if (shouldConsumePath) {
+            loadInitialUrl(
+                baseUrl = urlState.url,
+                path = path,
+                isNewServer = isNewServer,
+            )
+        } else {
+            // TODO not sure if we need to handle the path here, since if we end up into this
+            //  we would have already consumed the path and we are changing to another base URL
+            //  like going from internal to external.
```

Comment:
I would argue the path should be ignored in that case, as path is only for initial load of the activity, but any navigation after that will change it which we don't track.

Reply:
Yeah I share the same feeling but I wanted to discuss it here since I was not sure if it was really only for the first load.